File: mock-facility-management/src/main/java/de/vkb/dojo/es/facilityManagement/services/events/delegating/DelegatingEventAggregator.java
package de.vkb.dojo.es.facilityManagement.services.events.delegating;

import de.vkb.dojo.es.facilityManagement.model.event.Event;
import de.vkb.dojo.es.facilityManagement.services.events.EventAggregator;
import de.vkb.dojo.es.facilityManagement.services.events.EventAggregatorResult;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class DelegatingEventAggregator<E extends Event, A> implements EventAggregator<E, A> {
    private final List<PickyEventAggregator<E, ?, A>> aggregators = new ArrayList<>();

    public DelegatingEventAggregator(Collection<PickyEventAggregator<E, ?, A>> aggregators) {
        this.aggregators.addAll(aggregators);
    }

    public EventAggregatorResult<E, A> process(E event, A aggregate) {
        for (PickyEventAggregator<E, ?, A> aggregator : aggregators) {
            if (aggregator.canProcess(event)) {
                return aggregator.process(event, aggregate);
            }
        }
        throw new IllegalArgumentException("no aggregator responsible for event of class " + event.getClass().getName());
    }
}

File: mock-human-resources/src/main/kotlin/de/vkb/dojo/es/humanResources/services/events/delegating/DelegatingEventAggregator.kt
package de.vkb.dojo.es.humanResources.services.events.delegating

import de.vkb.dojo.es.humanResources.model.event.Event
import de.vkb.dojo.es.humanResources.services.events.EventAggregator
import de.vkb.dojo.es.humanResources.services.events.EventAggregatorResult


class DelegatingEventAggregator<E : Event, A>(aggregators: Collection<PickyEventAggregator<E, *, A>>?) :
    EventAggregator<E, A> {
    private val aggregators: MutableList<PickyEventAggregator<E, *, A>> = ArrayList()

    init {
        this.aggregators.addAll(aggregators!!)
    }

    override fun process(event: E, aggregate: A?): EventAggregatorResult<E, A> {
        for (aggregator in aggregators) {
            if (aggregator.canProcess(event)) {
                return aggregator.process(event, aggregate)
            }
        }
        throw IllegalArgumentException("no aggregator responsible for event of class " + event.javaClass.name)
    }
}
